### PR TITLE
New C++ Api: First batch of API guards.

### PR DESCRIPTION
**Pull request by aniemetz**
This is the first batch of API guards, mainly extending existing guards
in the Solver object with checks that Ops, Terms, Sorts, and datatype objects
are associated to this solver object.

This further changes how `DatatypeConstructorDecl` objects are created. Previously,
they were not created via the Solver object (while `DatatypeDecl` was). Now, they are
created via `Solver::mkDatatypeConstructorDecl`, consistent with how `DatatypeDecl`
objects are created.

### CODE REVIEW

**Comment by makaimann** [comment]
Also, FYI @kris-brown. This PR will change the datatypes interface.

File: src/api/cvc4cpp.cpp [hunk 1]
```diff
@@ -2749,13 +2775,27 @@ std::vector<Sort> Solver::mkDatatypeSortsInternal(
...
+  for (size_t i = 0, ndts = dtypedecls.size(); i < ndts; ++i)
+  {
+    CVC4_API_ARG_AT_INDEX_CHECK_EXPECTED(this == dtypedecls[i].d_solver,
+                                         "datatype declaration",
+                                         dtypedecls[i],
+                                         i)
+        << "a datatype declaration associated to this solver object";
+    CVC4_API_CHECK(this == dtypedecls[i].d_solver)
+        << "Given datatype declaration at index " << i
+        << " is not associated with this solver";
```
Comment (4tXJ7f) [inline src/api/cvc4cpp.cpp:2787]: It looks like these two checks are the same? Am I missing something?

Comment (aniemetz) [reply src/api/cvc4cpp.cpp:2787]: Indeed. The second was replaced by the first but I forgot to delete it.

**Comment by aniemetz** [comment]
Don't merge yet, we have a segfault in the datatypes Python example.

**Pull request merged**